### PR TITLE
Add unitsdata.cpp, getConversionRatesInfo, and unit tests.

### PR DESCRIPTION
**Pull request by hugovdm**
format/UnitsDataTest passes.

There are other tests within format/* that do fail: they're failing at current units-staging HEAD too though.

### CODE REVIEW

**Review by younies — APPROVED** [review]

**Comment by hugovdm** [comment]
Current status:
- This function is calculating the "base unit" because if I next want to fetch the unit preferences, I need to know what that base unit is.
- Because I'm calculating the base unit anyway, I check consistency of the base unit between "source" and "target" units. Which means I'm anyway checking convertability. This is why the function fails for unconvertable source/target.
- This function is *not* separating input and output conversion rates, and it is *not* combining the dimensionality information with each conversion rate.

Duplication concern:
- Next we need to caluclate more detailed conversion info, namely dimensionality of each input and output "single unit". Younies is working on this, in a new function, separate from mine. Which means it will essentually resplit and recalculate convertability etc.
- It may be reasonable to combine the two functions into one (the new one and the old one), such that we'll have one function that's similar to mine, but includes dimensionality info in the "ConversionRateInfo" struct.
- One could consider ripping out the base unit logic from my function as "unnecessary", however I need it for "getPreferenceData".
- In "getPreferenceData" (or whatever I called it) I was also fetching the preferences and the conversion rates for all those preferences, so also a bit of a blend. I may propose adjusting this function as I add that one.

I'm not convinced that returning a single vector is the best API: if we make it a void function and return the info via reference or pointer parameters, I could return separate info for source and target. And I could reuse the function to add additional preferences as part of implementing getPrerefenceData.

All that said, perhaps we just keep going and implement things as needed to get everything working correctly, and then afterwards we can consolidate, simplify, and optimise the code, merging functions as makes sense...

File: icu4c/source/i18n/unitsdata.h
Comment (younies) [inline icu4c/source/i18n/unitsdata.h]: Make the function take a list of MeasureUnits.

File: icu4c/source/i18n/unitsdata.h
Comment (younies) [inline icu4c/source/i18n/unitsdata.h]: This is not part of getting data function. We extract all the needed data first and then, we can have another function that checks some properties of the data.

For example, source and target can be reciprocal, However, your function will return an internal error.

**Review by younies — CHANGES_REQUESTED** [review]

File: icu4c/source/i18n/unitsdata.h
Comment (hugovdm) [reply icu4c/source/i18n/unitsdata.h]: I'm thinking it makes sense to take a single source and a list of targets - how does that sound?

Comment (hugovdm) [reply icu4c/source/i18n/unitsdata.h]: Switched to using a single MaybeStackVector. PTAL?

File: icu4c/source/i18n/unitsdata.h
Comment (hugovdm) [reply icu4c/source/i18n/unitsdata.h]: I've ripped it out, thanks. PTAL.

**Review by younies — APPROVED** [review]

File: icu4c/source/i18n/unitsdata.h
Comment (younies) [reply icu4c/source/i18n/unitsdata.h]: Done!

Thanks a lot

**Comment by sffc** [comment]
What's the difference between this PR and #32?  I reviewed #32 and I see a lot of the same code here.

**Comment by hugovdm** [comment]
This was the pull request I sent. In #32, Younies merged in the branch for this pull request, so by reviewing #32, you've reviewed two stacked changes.

We can discuss this in our meeting today.

**Comment by sffc** [comment]
I see, thanks for the clarification.  Why don't you fix this PR according to comments I left in the other PR, cleaning up the data sink to not have all the duplicated work.

File: icu4c/source/i18n/unitsdata.cpp
Comment (hugovdm) [inline icu4c/source/i18n/unitsdata.cpp]: Feedback from @sffc - the code we currently have (here in #31 and in #32 ) has a lot of "MaybeStackVector<ConversionRateInfo> &" being passed around. sffc suggests we make a first-class type called ConversionRateInfoList with the methods you need, that contains the MaybeStackVector inside.

To be reviewed together with #32's use-cases.

(This thread started at: https://github.com/sffc/icu/pull/32#discussion_r404695312 )

Comment (hugovdm) [reply icu4c/source/i18n/unitsdata.cpp]: My thoughts: I think returning a MaybeStackVector from getAllConversionRates might be reasonable enough. It could be packaged into a ConversionRateInfoList by other code? There is also a question of where to do the symbolic maths: but I think that might be best done when the converter is set up, if it's possible for the symbolic maths to cancel out? (Though if we don't care about maximum accuracy, we could simply deal with it in double precision maths.)

File: icu4c/source/i18n/unitsdata.cpp
Comment (hugovdm) [inline icu4c/source/i18n/unitsdata.cpp:94]: The sink is populating a MaybeStackVector it doesn't own. This seems to be a normal way of doing things - prior art review for a pointer or reference to the output data in a ResourceSink:

- ListFormatter::ListPatternsSink has a public char* read directly by its clients.
- RelDateFmtDataSink stores a pointer (`URelativeString *fDatesPtr;`),
- RelDateTimeFmtDataSink holds a reference passed to the constructor (`RelativeDateTimeCacheData &outputData;`).
- DecFmtSymDataSink holds a reference (`DecimalFormatSymbols& dfs;`)
- PluralTableSink holds a pointer (`UnicodeString *outArray;`)
- NumberRangeDataSink holds a reference (`NumberRangeData& fData;`), as does PluralRangesDataSink.
- ZNames::ZNamesLoader owns the loaded data and expects the client code to make a copy as needed.
- KeywordsSink (ucol_res.cpp) holds a pointer (`UList *values;`)

I went for a pointer, because stylistically I think it's clearer that there are "ownership concerns" (that one passes a pointer / an address is visible at the callsite).

(This new thread is a follow-on from https://github.com/sffc/icu/pull/32#discussion_r408212054 )

File: icu4c/source/i18n/unitsdata.cpp
Comment (sffc) [reply icu4c/source/i18n/unitsdata.cpp]: The typical pattern is that you heap-allocate an object to store the results from a data loading operation, and use the sink to populate it.  I think you should do the same here.  I would rather err on the side of too many types than too few; I often add types where I didn't have types before, but very rarely the other direction.

For the time being, if it's easier, you can add a TODO in order to unblock #32.

Comment (hugovdm) [reply icu4c/source/i18n/unitsdata.cpp]: I first switched to passing in a MaybeStackVector as a pointer parameter, on the stylistic choice that it's clearer at the call-site that it's an output parameter.

However these (naive) grep's suggest references are more common in ICU:

    egrep '& result|&result' i18n/*.h
    egrep '\* result|\*result' i18n/*.h

so I've now changed it to references. PTAL?

File: icu4c/source/i18n/unitsdata.cpp
Comment (hugovdm) [inline icu4c/source/i18n/unitsdata.cpp:79]: This line is introducing an `operator new(unsigned long)` symbol, with the healthy-code saying we shouldn't (via cmemory.h:728, MemoryPool code):

    MaybeStackVector<ConversionRateInfo>::emplaceBack();

By comparison, with measunit_extra.cpp, we're getting a `icu_66::UMemory::operator new(unsigned long)`:

    MaybeStackVector<SingleUnitImpl>::emplaceBack();

Looking into header files and other context to figure out how this is resolving differently...

Comment (hugovdm) [reply icu4c/source/i18n/unitsdata.cpp:79]: struct SingleUnitImpl : public UMemory { ... }

**Comment by hugovdm** [comment]
Is this good to merge?

File: icu4c/source/i18n/unitsdata.cpp
Comment (sffc) [reply icu4c/source/i18n/unitsdata.cpp]: I would still rather see an extra type ConversionRateInfoList encapsulating the MaybeStackVector, but not strongly enough to block this PR.

**Review by sffc — APPROVED** [review]

**Pull request merged**
